kubernetes-model/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java

```diff
@@ -152,42 +111,133 @@ public static void registerCustomKind(String kind, Class<? extends KubernetesRes
-        if (result != null) {
-          return result;
+        private synchronized Map<String, Class<? extends KubernetesResource>> getMappings() {
+            if (mappings == null) {
+                this.mappings = new HashMap<>();
+                getAllMappingProviders().forEach(this::registerProvider);
+            }
+            return mappings;
         }
-      }
 
-        return null;
-    } 
+        protected Stream<KubernetesResourceMappingProvider> getAllMappingProviders() {
+            //Use service loader to load extension types.
+            Iterable<KubernetesResourceMappingProvider> currentThreadClassLoader =
+                    () -> ServiceLoader.load(KubernetesResourceMappingProvider.class, Thread.currentThread().getContextClassLoader())
+                            .iterator();
+            Iterable<KubernetesResourceMappingProvider> classClassLoader =
+                    () -> ServiceLoader.load(KubernetesResourceMappingProvider.class, KubernetesDeserializer.class.getClassLoader())
+                            .iterator();
+            return Stream.concat(
+                    StreamSupport.stream(currentThreadClassLoader.spliterator(), false),
+                    StreamSupport.stream(classClassLoader.spliterator(), false))
+                    .filter(distinctByClassName(KubernetesResourceMappingProvider::getClass));
+        }
 
-    private static Class loadClassIfExists(String className) {
-        try {
-            return KubernetesDeserializer.class.getClassLoader().loadClass(className);
-        } catch (Throwable t) {
+        private String getClassName(String key) {
+            if (key != null && key.contains(KEY_SEPARATOR)) {
+                return key.substring(key.indexOf(KEY_SEPARATOR) + 1);
+            } else {
+                return key;
+            }
+        }
+
+        private Class<? extends KubernetesResource> getInternalTypeForName(String name) {
+            for (String aPackage : PACKAGES) {
+                Class<? extends KubernetesResource> result = loadClassIfExists(aPackage + name);
+                if (result != null) {
+                    return result;
+                }
+            }
             return null;
         }
+
+        private Class<? extends KubernetesResource> loadClassIfExists(String className) {
+            try {
+                Class<?> clazz = KubernetesDeserializer.class.getClassLoader().loadClass(className);
+                if (!KubernetesResource.class.isAssignableFrom(clazz)) {
+                    return null;
+                }
+                return (Class<? extends KubernetesResource>) clazz;
+            } catch (Throwable t) {
+                return null;
+            }
+        }
+
+        private Predicate<KubernetesResourceMappingProvider> distinctByClassName(
+                Function<KubernetesResourceMappingProvider, Class<? extends KubernetesResourceMappingProvider>> mapperProvider) {
```

Comment:
could you take care of the indentation here?

Reply:
@dev-gaur this identation is on purpose, the whole line is too long to fit in 120 characters. Spliting parts isn't readble since it's a single parameter with lengthy generdics declaration. I therefore opted to keep the param all together but pushed it onto it's own line that that's indentated by 4. Makes sense?